1092 review edition 2 rl ii (#1094)
geocompx/geocompr: 01-introduction.Rmd

Geocomputation\index{geocomputation} implies working with geographic data in a reproducible code-driven environment and programming new results, methods and tools, which is what this book is all about.\index{GIS!connotations}

## Why open source tools for geocomputation? {#why-open-source}

Early geographers used a variety of tools including barometers, compasses and [sextants](https://en.wikipedia.org/wiki/Sextant) to advance knowledge about the world [@wulf_invention_2015]. 
It was only with the invention of the marine [chronometer](https://en.wikipedia.org/wiki/Marine_chronometer) in 1761 that it became possible to calculate longitude at sea, enabling ships to take more direct routes, for example.
Using a sextant was a difficult-to-learn but essential tool for navigation in the 18th century.
Likewise, learning to program and write and share reproducible code can be hard, but is a key skill for modern geocomputation.
<!-- Maps were predominantly hand-drawn until modern computing enabled digitisation, a process which only become widespread in the 1990s and which is still ongoing [@gold_outsidein_1996; @auffret_histmapr_2017]. -->

> Nowadays, researchers and practitioners have no such limitations and in some cases face the opposite problem: too much data; too many tools.

Reviewer: This sentence have a very week connection with the previous one...

Author: Agreed.

Author: Fixed by adding this to previous paragraph: Before the turn of the century, there was an accute shortage of data and tools for geographic analysis.

Most phones now have a global positioning (GPS\index{GPS}) receiver.

> Sensors ranging from satellites and semi-autonomous vehicles to citizen scientists incessantly measure every part of the world.

Reviewer: Are citizen scientists "sensors"?

Author: I think so in this context.

With the rapidly changing hardware, software and data landscapes, it's important to choose tools that are future-proof, able to adapt.
A major advantage of open source software is its **rate of development and longevity**, with thousands of potential contributors.
Hundreds of people submit bug reports, suggested new features, and documentation improvements to open source projects every day, a rate of evolution that most proprietary solutions simply cannot keep up with.

Another major advantage of open source software is its **interoperability**.
While proprietary products tend to be monolithic 'empires' that are difficult to maintain (linked to the previously mentioned advantage), open source software is more like a 'federation' of modular tools that can be combined in different ways.
This has allowed open source data science languages such as R to rapidly incorporate new developments such as interfaces to high performance visualisation libraries and file formats, while proprietary solutions struggle to keep up.

> Another major advantage is **reproducibility**.

Reviewer: This paragraph starts exactly the same as the previous one...

Author: Woops. Thanks Jakub, will fix.

Author: Updated sentences fix it:

A linked advantage is **interoperability**.
While proprietary products tend to be monolithic 'empires' that are difficult to maintain (linked to the previously mentioned advantage), open source software is more like a 'federation' of modular tools that can be combined in different ways.
This has allowed open source data science languages such as R to rapidly incorporate new developments such as interfaces to high performance visualisation libraries and file formats, while proprietary solutions struggle to keep up.

Another major advantage is **reproducibility**.
Being able to replicate findings is vital for scientific research, and open source software removes an important potential barrier to others being able to check your findings or apply your methods in new places.
The combination of using tools that can be accessed by anyone for free with the ability to share code and data means that the results of your work can be checked and built upon by others, a huge advantage if you want your work to be used and cited.

Likewise, learning to program and write and share reproducible code can be hard, but is a key skill for modern geocomputation. -->
<!-- Maps were predominantly hand-drawn until modern computing enabled digitisation, a process which only become widespread in the 1990s and which is still ongoing [@gold_outsidein_1996; @auffret_histmapr_2017]. -->

Nowadays, researchers and practitioners have no such limitations and in some cases face the opposite problem: too much data; too many tools.
Most phones now have a global positioning (GPS\index{GPS}) receiver.
Sensors ranging from satellites and semi-autonomous vehicles to citizen scientists incessantly measure every part of the world.
The rate of data produced can be overwhelming, with emerging technologies such as autonomous vehicles generating hundreds or even thousands of GB of data daily.
Remote sensing\index{remote sensing} datasets from satellites are too large to analyze with a single computer, as outlined in Chapter \@ref(gis).
This 'geodata revolution' drives demand for high performance computer hardware and efficient, scalable software to handle and extract signal from the noise.

> Evolving open source tools can import and process subsets from the vast geographic data stores directly, and via APIs and via interfaces to databases. \index{spatial database}

Reviewer 2: Something is missing in this sentence. Do you mean that we can use APIs for an easy cli-driven data access and other interfaces for interacting with databases (import, manipulation and export again)?

Author: Will look this eve, thanks for the review!

> With the rapidly changing hardware, software and data landscapes, it's important to choose tools that are future-proof, able to adapt.

Reviewer 2: and able to adapt to frequent changes, new needs, new technologies?

Author: ```suggestion
With the rapidly changing hardware, software and data landscapes, it's important to choose tools that are future-proof.
```

A major advantage of open source software is its **rate of development and longevity**, with thousands of potential contributors.
Hundreds of people submit bug reports, suggested new features, and documentation improvements to open source projects every day, a rate of evolution that most proprietary solutions simply cannot keep up with.

A linked advantage is **interoperability**.
While proprietary products tend to be monolithic 'empires' that are difficult to maintain (linked to the previously mentioned advantage), open source software is more like a 'federation' of modular tools that can be combined in different ways.
This has allowed open source data science languages such as R to rapidly incorporate new developments such as interfaces to high performance visualisation libraries and file formats, while proprietary solutions struggle to keep up.

Another major advantage is **reproducibility**.

> Being able to replicate findings is vital for scientific research, and open source software removes an important potential barrier to others being able to check your findings or apply your methods in new places.

Reviewer 2: ```suggestion
Being able to replicate findings is vital for scientific research, and open source software removes an important barrier of reproducibility by enabling others to check your findings or applying your methods in new contexts using the same tools.
```

# cranlogs::cran_top_downloads(when = "last-month") # most downloaded pkgs
```

> A parallel group of developments relates to the [rspatial](https://github.com/rspatial) set of packages.^[Note the difference between "r-spatial", organization containing packages such as **sf**, and "rspatial", organization responsible for **terra**.]

Reviewer: @Robinlovelace maybe it would be good to keep the footnote about the rspatial/r-spatial difference

Author: Agree it's potentially useful info. Can be found out online and not sure it lives in the book. No big objections to re-adding it or strong feelings, just felt a bit like too much info to me and footnotes are fiddly and generally discouraged I think.

Thoughts @jannes-m to get another opinion?

Some of these, notably **spatial**, **sgeostat** and **splancs** are still available on CRAN\index{CRAN} [@rowlingson_splancs_1993; @rowlingson_splancs_2017;@venables_modern_2002; @majure_sgeostat_2016].
Key spatial packages were described in @ripley_spatial_2001, which outlined R packages for spatial smoothing and interpolation [@akima_akima_2016; @jr_geor_2016] and point pattern analysis [@rowlingson_splancs_2017; @baddeley_spatial_2015].
One of these (**spatstat**) is still being actively, more than 20 years after it was first released.

A following commentary outlined the future prospects of spatial statistics [@bivand_more_2001], setting the stage for the development of the popular **spdep** package [@bivand_spdep_2017].
Notably, the commentary mentioned the need for standardization of spatial interfaces, efficient mechanisms for exchanging data with GIS\index{GIS}, and handling of spatial metadata such as coordinate reference systems (CRS\index{CRS}), aims which have largely been achieved.

> **maptools** [@bivand_maptools_2017] is another important package from this time, which provided an interface to the [shapelib](http://shapelib.maptools.org/) library for reading the Shapefile\index{Shapefile} file format and which fed into the **sp** package on which **sf** is based.

Reviewer: I suggest not using the "on which **sf** is based." part (it can be wrongly understood)

Author: :+1: not added yet. Suggestion (allowing anyone to add I think):

```suggestion
**maptools** [@bivand_maptools_2017] is another important package from this time, which provided an interface to the [shapelib](http://shapelib.maptools.org/) library for reading the Shapefile\index{Shapefile} file format and which fed into **sp**.
```

In parallel with these developments of class systems and methods came the support for R as an interface to dedicated GIS software.
**GRASS** [@bivand_using_2000] and follow-on packages **spgrass6**, **rgrass7** and **rgrass** (for GRASS GIS\index{GRASS GIS} GIS 6, 7, 7 + 8 respectively) were prominent examples in this direction [@bivand_rgrass7_2016;@bivand_spgrass6_2016;@R-rgrass].

> Other examples of bridges between R and GIS include bridges to QGIS via **qgisprocess** [@R-qgisprocess], SAGA via **Rsagacmd** [@R-Rsagacmd] or **RSAGA** [@R-RSAGA]\index{RSAGA (package)} and ArcGIS via **RPyGeo** [@brenning_arcgis_2012, first published in 2008], and more (see Chapter \@ref(gis)).

Author: :+1:

Since the publication of the first version of Geocomputation with R in 2018, several packages for spatial data visualization have been developed and improved.
The **rayshader** package, for example, enables the development of striking and easy-to-animate 3D visualisations via raytracing and multiple hill-shading methods [@morganwall_rayshader_2021].
\index{ggplot2 (package)}
The very popular **ggplot2** package gained new spatial capabilities, thanks to work on the **ggspatial** package, which provides scale bars and north arrows [@dunnington_ggspatial_2021].
**gganimate** enables smooth and customizable spatial animations [@pedersen_gganimate_2020].

Existing visualization packages have also been improved or re-written.
Large raster objects are automatically downscaled in **tmap** and high-performance interactive maps are now possible thanks to packages including **leafgl** and **mapdeck**.
<!-- TODO: add release date of tmap 4.0 when ready -->
The **mapsf** package (successor of **cartography**) was re-written to reduce dependencies and improve performance [@giraud_mapsf_2021]; and **tmap** underwent a major update in Version 4, in which most of the internal code was revised.

> Version 4 of the **tmap** package, in which most of the internal code was revised.

Reviewer: @Robinlovelace repeated sentence

Author: ```suggestion
```

> In late 2021, the planned retirement of **rgdal**, **rgeos** and **maptools** [was announced](https://stat.ethz.ch/pipermail/r-sig-geo/2021-September/028760.html) and in October 2023 they were removed from CRAN.

Reviewer: "archived on" instead of "removed from"?
